Add pagination

kul3r4 opened the pull request:
Solves #430

sebastianbenz commented on backend/amp-live-list.go:
```diff
@@ -157,13 +163,56 @@ func createMetadata(r *http.Request) []BlogPosting {
 	blogItems := getBlogEntries(newStatus, timestamp)
 	score := createScore(newStatus, 0)
-	return LiveBlogSample{BlogItems: blogItems, FootballScore: score, BlogMetadata: createMetadata(r)}
+	firstItemIndex, _ := getProductIndexFromID(firstBlogID, blogItems)
```
don't ignore the error?

kul3r4 replied:
Maybe I should remove the error, if the item cannot be found at the index provided, it should simply show the first page. What do you think?

sebastianbenz replied:
sgtm

sebastianbenz commented on src/50_Samples_%26_Templates/Live_Blog.html:
```diff
@@ -173,11 +187,19 @@
     </div>
     <!-- #### Blog Posts -->
     <!-- Use amp-live-list to implement a live blog. Find an amp-live-list example [here](/components/amp-live-list) -->
```
Can you explain the pagination feature? Or do you plan to add this to the amp-live-list sample?

sebastianbenz commented on src/50_Samples_%26_Templates/Live_Blog.html:
```diff
@@ -200,6 +222,21 @@ <h4 class="title">[[.Heading]]</h4>
+    <div pagination>
+      <nav aria-label="amp live list pagination">
+      <ul class="pagination">
+        [[if eq .PrevPageURL ""]][[else]]
```
```
[[if eq .PrevPageURL ""]][[else]]
 +          <li><a href="[[.PrevPageURL]]">Prev</a></li>
 +        [[end]]
```
can be replaced with:
```
[[if .PrevPageURL ]]
       <li><a href="[[.PrevPageURL]]">Prev</a></li>
 [[end]]
```

sebastianbenz commented on src/50_Samples_%26_Templates/Live_Blog.html:
```diff
@@ -173,11 +187,19 @@
     </div>
     <!-- #### Blog Posts -->
     <!-- Use amp-live-list to implement a live blog. Find an amp-live-list example [here](/components/amp-live-list) -->
+    [[if eq .PrevPageURL ""]]
```
use instead:
 
```
[[if .PrevPageURL]] 

[[else]]
...
```

sebastianbenz commented on backend/amp-live-list.go:
```diff
@@ -157,13 +163,56 @@ func createMetadata(r *http.Request) []BlogPosting {
+	return (index / MAX_BLOG_NUMBER) + 1
+}
+
+func getProductIndexFromID(id string, blogItems []BlogItem) (int, error) {
```
Why not store the blogs in an array and use the index as ID?

sebastianbenz commented on backend/amp-live-list.go:
```diff
@@ -157,13 +163,56 @@ func createMetadata(r *http.Request) []BlogPosting {
+func getNextPageUrl(r *http.Request, page Page, blogItems []BlogItem, nextPageFirstItemIndex int) string {
+	if nextPageFirstItemIndex < len(blogItems) {
+		nextPageFirstItemID := blogs[nextPageFirstItemIndex].ID
+		return buildSourceOrigin(r.Host) + page.Route + "?from=" + nextPageFirstItemID
```
you build the same URL in both methods getNextPageUrl and getPrevPageUrl. Make this method only return the new index and build the URL somewhere else.

sebastianbenz commented on backend/amp-live-list.go:
```diff
@@ -23,6 +25,7 @@ import (
 const (
 	AMP_LIVE_LIST_COOKIE_NAME = "ABE_AMP_LIVE_LIST_STATUS"
 	FIFTEEN_SECONDS           = 15
+	MAX_BLOG_NUMBER           = 5
```
What does MAX_BLOG_NUMBER mean? Can you find a better name?

kul3r4 commented:
@sebastianbenz PTAL

sebastianbenz commented on backend/amp-live-list.go:
```diff
@@ -146,7 +154,7 @@ func createMetadata(r *http.Request) []BlogPosting {
 		result = append(result, BlogPosting{"BlogPosting",
 			blogs[i].Heading,
-			buildSourceOrigin(r.Host) + "/" + CATEGORY_SAMPLE_TEMPLATES + "/live_blog/#" + blogs[i].ID,
+			buildSourceOrigin(r.Host) + "/" + CATEGORY_SAMPLE_TEMPLATES + "/live_blog/#" + fmt.Sprintf("%s%d", BLOG_ID_PREFIX, i+1),
```
I find it weird to combine concatenation via `+` and `fmt.Sprintf`. Please use only one.

sebastianbenz replied:
Also, `buildSourceOrigin(r.Host) + "/" + CATEGORY_SAMPLE_TEMPLATES + "/live_blog/#` should be done outside the loop and not repeated every iteration.

sebastianbenz commented on backend/amp-live-list.go:
```diff
@@ -157,13 +165,71 @@ func createMetadata(r *http.Request) []BlogPosting {
+	nextPageFirstItemIndex := firstItemIndex + MAX_BLOG_ITEMS_NUMBER_PER_PAGE
+	nextPageId := getNextPageId(r, page, blogItems, nextPageFirstItemIndex)
+	if nextPageId != "" {
+		return buildPaginationURL(r, page, nextPageId)
```
That's not what I meant - this method shouldn't build the URL, build the URL before calling this method. This method should simply calculate the index.

kul3r4 replied:
This is an helper method I added.
we have now:
getNextPageId -> which only builds the id
buildPaginationURL -> which builds the URL
getNextPageUrl -> which combines the 2 above
Should I remove the last one?

sebastianbenz replied:
You build the URL twice instead of:

```
urlPrefix := buildURL(...)
nextIndex := nextId(...)
previousIndex := prevId(...)
nextUrl := urlPrefix + nextIndex
prevUrl := urlPrefix + prevIndex
```

kul3r4 replied:
I see it now, thanks! I pushed some changes,
PTAL

sebastianbenz commented on backend/amp-live-list.go:
```diff
@@ -157,13 +165,71 @@ func createMetadata(r *http.Request) []BlogPosting {
+	blogEntryId, error := strconv.Atoi(id)
+	//default to the first page
+	if error != nil || blogEntryId > len(blogItems) {
+		return 0
```
why return 0 if they start from 1 according to the following comment?

kul3r4 replied:
Because I am returning the index, not the id.
So for item 1, the ID would be post1 and the index would be 0.

sebastianbenz commented on src/50_Samples_%26_Templates/Live_Blog.html:
```diff
@@ -200,6 +221,21 @@ <h4 class="title">[[.Heading]]</h4>
+    <div pagination>
+      <nav aria-label="amp live list pagination">
+      <ul class="pagination">
+        [[if.PrevPageURL]]
```
nit: s/if.PrevPageURL/if .PrevPageURL/

sebastianbenz commented on backend/amp-live-list.go:
```diff
@@ -157,13 +165,71 @@ func createMetadata(r *http.Request) []BlogPosting {
 	blogItems := getBlogEntries(newStatus, timestamp)
 	score := createScore(newStatus, 0)
-	return LiveBlogSample{BlogItems: blogItems, FootballScore: score, BlogMetadata: createMetadata(r)}
+	firstBlogID := strings.TrimPrefix(r.URL.Query().Get("from"), BLOG_ID_PREFIX)
```
perform this `strings.TrimPrefix(r.URL.Query().Get("from")` outside this method. Creating a blog entry should not depend on the request.

sebastianbenz commented:
Can you fix the build?

sebastianbenz commented on src/50_Samples_%26_Templates/Live_Blog.html:
```diff
@@ -172,12 +185,20 @@
     </div>
     <!-- #### Blog Posts -->
-    <!-- Use amp-live-list to implement a live blog. Find an amp-live-list example [here](/components/amp-live-list) -->
+    <!-- Use amp-live-list to implement a live blog. Find an amp-live-list example [here](/components/amp-live-list). Use `<div pagination>` tag to include pagination, find more [here](https://www.ampproject.org/docs/reference/components/amp-live-list) -->
```
sug: "Use `<div pagination>` tag to include pagination, find more [here](https://www.ampproject.org/docs/reference/components/amp-live-list)" -> "You can implement pagination for live-blogs using [`<div pagination>`](https://www.ampproject.org/docs/reference/components/amp-live-list)

sebastianbenz commented on spec/compiler/CodeSectionSpec.js:
```diff
@@ -160,6 +160,12 @@ describe("CodeSection", function() {
       expect(section.hideColumns()).toEqual(true);
     });
   });
+  describe('removes uncorrectly escaped templates', function(){
```
Can you please cover the most common cases where we have to clean the code.

sebastianbenz commented on tasks/validate-example.js:
```diff
@@ -40,7 +40,7 @@ module.exports = function() {
       if (file.metadata &&
         (file.metadata.experiments || file.metadata.skipValidation) ||
-         !file.path.endsWith('.html')) {
+         !file.path.endsWith('.html') || file.path.includes('live_blog')) {
```
please don't hardcode samples - use metadata instead. If you look at the code you will see that there is already a flag for skipping sample validation...

kul3r4 commented:
@sebastianbenz I added 2 more tests.
I also fixed the spaces for the json, have a look: [here](https://chiara-ampbyexample.appspot.com/samples_templates/live_blog/)
@jkingyens Would you mind taking a look at https://chiara-ampbyexample.appspot.com/samples_templates/live_blog/ from the point of view of metadata?

sebastianbenz commented on spec/compiler/CodeSectionSpec.js:
```diff
@@ -160,6 +160,18 @@ describe("CodeSection", function() {
+      expect(section.cleanUpCode("<span class=\"hljs-string\">\"[[.Timestamp]]\"</span>")).toEqual("<span class=\"hljs-string\">\"[[.Timestamp]]\"</span>")
+    });
+    it('does not alter valid escaped templates which contains if clause', function() {
+      expect(section.cleanUpCode("</span>[[if .PrevPageURL]]<span class=\"hljs-tag\">")).toEqual("</span>[[if .PrevPageURL]]<span class=\"hljs-tag\">")   
```
Not sure this test makes sense. Much more interesting would be whether  escaped ifs, range & co work.

sebastianbenz commented on src/50_Samples_%26_Templates/Live_Blog.html:
```diff
@@ -172,12 +186,14 @@
     <!-- #### Blog Posts -->
-    <!-- Use amp-live-list to implement a live blog. Find an amp-live-list example [here](/components/amp-live-list) -->
-    <amp-live-list
+    <!-- Use amp-live-list to implement a live blog. The amp-live-list component regularly polls the host document for updated content and updates the end user's browser as new items become available. This means that every time a new post needs to be added, the host document should be updated by the CMS to include the update both the body and the metadata section. Find an amp-live-list example [h ...
```
Can you please add a line break after this paragraph.

sebastianbenz commented on src/50_Samples_%26_Templates/Live_Blog.html:
```diff
@@ -172,12 +186,14 @@
-    <!-- Use amp-live-list to implement a live blog. Find an amp-live-list example [here](/components/amp-live-list) -->
-    <amp-live-list
+    <!-- Use amp-live-list to implement a live blog. The amp-live-list component regularly polls the host document for updated content and updates the end user's browser as new items become available. This means that every time a new post needs to be added, the host document should be updated by the CMS to include the update both the body and the metadata section. Find an amp-live-list example [h ...
+    Long blogs could use pagination to improve performance by limiting the number of blog items to be displayed on a page. To implement pagination, amp-live-list allows to add the `<div pagination></div>` element, to then insert any markup needed for pagination, for example the page number, a link to the next and previous page. -->
```
Perfect! Thanks!

kul3r4 commented:
@sebastianbenz PTAL

sebastianbenz commented on tasks/lib/CodeSection.js:
```diff
@@ -133,5 +132,10 @@ module.exports = class CodeSection {
   }
+
+  cleanUpCode(input) {
+    const encodedTemplateRegexp = /\[\[\s*\n*<span\sclass="hljs\-attr">(range|if|else|end)?(<\/span><span\sclass="hljs\-attr">)?(\.[A-Za-z]*)?<\/span>\s*\n*\]\]/g
```
Can you move the regex declaration to a constant at the top of the file. 

* `/s*/n*` -> the /n will never match as /s matches all whitespaces
* can you simplify `<span\sclass="hljs\-attr">` to `<.*?>` ? (I'm not sure)
* can you simplify `(range|if|else|end)?` to `(.)*?` ? (I'm not sure)

sebastianbenz commented on spec/compiler/CodeSectionSpec.js:
```diff
@@ -160,6 +160,24 @@ describe("CodeSection", function() {
+      expect(section.cleanUpCode("[[<span class=\"hljs-attr\">if</span><span class=\"hljs-attr\">.BlogItems</span>]]")).toEqual("[[if .BlogItems]]")   
+    });
+    it('contains an escaped template with end clause', function() {
+      expect(section.cleanUpCode("[[<span class=\"hljs-attr\">end</span>]]")).toEqual("[[end ]]")   
```
What happens if templates contain whitespace? 

```
[[  if   .Test   ]]
[[  .Test   ]]
...
```

kul3r4 commented:
@sebastianbenz PTAL

sebastianbenz commented on src/20_Components/amp-live-list.html:
```diff
@@ -75,7 +75,7 @@
-    This sample shows an implementation with a server rendering a series of 10 elements. `amp-live-list` will poll the page every 15 seconds. Every time `amp-live-list` polls the page, the server adds a new item to the page. Refer to [server.go](https://github.com/ampproject/amp-by-example/blob/master/server.go) and [amp-live-list.go](https://github.com/ampproject/amp-by-example/blob/master/backe ...
+    This sample shows an implementation with a server rendering a series of 5 elements. `amp-live-list` will poll the page every 15 seconds. Every time `amp-live-list` polls the page, the server adds a new item to the page. Refer to [server.go](https://github.com/ampproject/amp-by-example/blob/master/server.go) and [amp-live-list.go](https://github.com/ampproject/amp-by-example/blob/master/backen ...
 
     This sample stops after ten updates. If you want to start over, delete cookies from ampbyexample.com or open this page again in an incognito/private tab.
```
Is this still correct? Doesn't it stop after 5 items?
